src/tests/test_particle.c

```diff
@@ -0,0 +1,754 @@
+      cgio_cleanup();
+      exit(1);
+   }
+   }
+}
+
+/* This tests MLL functions that handle reading and writing of core particle nodes i.e. ParticleZone_t,
+ * ParticleCoordinates_t, ParticleSolution_t, ParticleIterativeData_t and particle fields i.e. DataArray_t */
+static void test_particle_io_main()
+{
+    int fnum, bnum, znum, snum, cnum;
+    char *outfile = "test_particle.cgns";
+
+    unlink (outfile);
+    printf ("creating CGNS file %s\n", outfile);
+
+    cg_configure(CG_CONFIG_ERROR, (void*) print_error);
+
+    cg_open (outfile, CG_MODE_WRITE, &fnum);
+    cg_base_write (fnum, "Base", 3, 3, &bnum);
+    cg_goto(fnum, bnum, NULL);
+    cg_dataclass_write(CGNS_ENUMV(NormalizedByDimensional));
+
+    /* Write BaseIterativeData_t so that we can write out ParticleIterativeData_t later on */
+    double output_time[1] = {0};
+    double iteration[1] = {1};
+    cgsize_t length = 1;
+    cg_goto(fnum, bnum, "end");
+    cg_biter_write(fnum, bnum, "BaseIterativeData", 1);
+    cg_gopath(fnum, "./BaseIterativeData");
+    cg_array_write("TimeValues", RealDouble, 1, &length, output_time);
+    cg_array_write("IterationValues", Integer, 1, &length, iteration);
+
+    float exp[5];
+    for (int n = 0; n < 5; n++)
+        exp[n] = (float)0.0;
+    exp[1] = (float)1.0;
+
+    cgsize_t zone_size[3];
+    zone_size[0] = 11;
+    zone_size[1] = 4;
+    zone_size[2] = 0;
+
+    cg_zone_write (fnum, bnum, "Cells", zone_size, Unstructured, &znum);
+    cg_coord_write (fnum, bnum, znum, RealDouble, "CoordinateX", xc, &cnum);
+    cg_goto(fnum, bnum, "Zone_t", znum, "GridCoordinates", 0, "CoordinateX", 0, NULL);
+    cg_exponents_write(CGNS_ENUMV(RealSingle), exp);
+    cg_coord_write (fnum, bnum, znum, CGNS_ENUMV(RealDouble), "CoordinateY", yc, &cnum);
+    cg_gopath(fnum, "../CoordinateY");
+    cg_exponents_write(CGNS_ENUMV(RealSingle), exp);
+    cg_coord_write (fnum, bnum, znum, CGNS_ENUMV(RealDouble), "CoordinateZ", zc, &cnum);
+    cg_gopath(fnum, "../CoordinateZ");
+    cg_exponents_write(CGNS_ENUMV(RealSingle), exp);
+
+    cgsize_t ne = 0;
+
+    /* NGON_n first so polyhedra face references are correct */
+    cg_poly_section_write (fnum, bnum, znum, "NGON_n", CGNS_ENUMV(NGON_n), ne+1, ne+npoly, 0, poly, poly_offsets, &snum);
+    ne += npoly;
+
+    /* NFACE_n */
+    cg_poly_section_write (fnum, bnum, znum, "NFACE_n", CGNS_ENUMV(NFACE_n), ne+1, ne+nface, 0, face, face_offsets, &snum);
+
+    int pnum, pcnum;
+
+    char particle_names[3][32];
+    strcpy(particle_names[0], "IsoOctane");
+    strcpy(particle_names[1], "Oil");
+    strcpy(particle_names[2], "Water");
+
+    cgsize_t particle_sizes[3] = {0, 0, 2};
+
+    /* Create ParticleZone_t nodes */
+    /* It is okay to define ParticleZone_t nodes with 0 particles
+     * this implies that for the current solution time, no particles
+     * of said type (IsoOctane and Oil) are in the Base */
+    cg_particle_write(fnum, bnum, "IsoOctane", 0, &pnum); // 0 IsoOctane particles
```

Comment:
Can you add pass-fail checks for the new particle APIs?

Reply:
@brtnfld By pass-fail checks, did you mean that I should use write_test_status (similar to pcgns_ctest.c)?

Reply:
I meant to check the API's return status to ensure it did not fail.

Reply:
Those should get passed to the error call back function (print_error) and printed to the screen right? We have 1 test in that file where we expect an error (trying to create a particle bbox with ints, but this shouldn't be fatal). Fatal errors in any API function i.e. code -1 will cause the test to fail (exit(1)).  I can remove this call back and put all function inside an if statement if that is preferred.

Reply:
I'm sorry; I missed the cg_configure. Yes, a callback is OK. However, that will not work with the test status schema unless you print the test header and then print FAIL in the callback. I don't know what that will gain except that it would indicate at what point in the test it fails. It is fine as is.